### PR TITLE
Speed up Settings Filtering

### PR DESCRIPTION
**Pull request by original-brownbear**
Settings filtering can make use of the fact that we're working with a prefix tree in a number of cases.
Also, we can avoid a lot of redundant empty settings instances here and there and have faster `exists`
checks by using a hash-set instead of a tree-set even if there's no secure settings.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/settings/Settings.java [hunk 1]
```diff
@@ -1271,24 +1299,33 @@ private FilteredMap(Map<String, Object> delegate, Predicate<String> filter, Stri
             this.delegate = delegate;
             this.filter = filter;
             this.prefix = prefix;
+            if (filter == null) {
+                size = delegate.size();
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/common/settings/Settings.java:1303]: nit: use this.size for assignment, for consistency

File: server/src/main/java/org/elasticsearch/common/settings/Settings.java [hunk 2]
```diff
@@ -198,15 +207,22 @@ private Object convertMapsToArrays(Map<String, Object> map) {
      * A settings that are filtered (and key is removed) with the specified prefix.
      */
     public Settings getByPrefix(String prefix) {
-        return new Settings(new FilteredMap(this.settings, (k) -> k.startsWith(prefix), prefix), secureSettings == null ? null :
-            new PrefixedSecureSettings(secureSettings, prefix, s -> s.startsWith(prefix)));
+        if (prefix.isEmpty()) {
+            return this;
+        }
+        char[] toPrefixCharArr = prefix.toCharArray();
+        toPrefixCharArr[toPrefixCharArr.length - 1]++;
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/common/settings/Settings.java:214]: It took me a little bit thinking about how this works. A short comment would be appreciated, explaining this adjusts the last character of the prefix for the ending key that is non-inclusive.

**Review by rjernst — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Thanks Ryan!

**Pull request merged**